File: NtgeCore-Android/src/main/java/com/dimension/ntge/ed25519/Ed25519Keypair.kt
package com.dimension.ntge.ed25519

import com.dimension.ntge.Ntge

class Ed25519Keypair internal constructor(
        internal val ptr: Long
) : AutoCloseable {

    companion object {
        fun new(): Ed25519Keypair {
            return Ed25519Keypair(Ntge.newEd25519Keypair())
        }
        fun fromPrivateKey(privateKey: Ed25519PrivateKey): Ed25519Keypair {
            return Ed25519Keypair(Ntge.getEd25519KeypairFromPrivateKey(privateKey.ptr))
        }
    }

    val privateKey = Ed25519PrivateKey(Ntge.getPrivateKeyFromEd25519Keypair(ptr))
    val publicKey = Ed25519PublicKey(Ntge.getPublicKeyFromEd25519Keypair(ptr))

    override fun close() {
        Ntge.destroyEd25519Keypair(ptr)
        privateKey.close()
        publicKey.close()
    }
}

File: NtgeCore.Net/src/Ed25519/Ed25519Keypair.cs
using System;

namespace NtgeCore.Net.Ed25519
{
    public class Ed25519Keypair : RustObject
    {
        private Ed25519PrivateKey? _privateKey;
        private Ed25519PublicKey? _publicKey;

        internal Ed25519Keypair(IntPtr ptr) : base(ptr)
        {
        }

        public Ed25519PrivateKey PrivateKey
        {
            get
            {
                if (_privateKey == null)
                {
                    _privateKey = new Ed25519PrivateKey(Native.getPrivateKeyFromEd25519Keypair(Ptr));
                }
                return _privateKey;
            }
        }

        public Ed25519PublicKey PublicKey
        {
            get
            {
                if (_publicKey == null)
                {
                    _publicKey = new Ed25519PublicKey(Native.getPublicKeyFromEd25519Keypair(Ptr));
                }
                return _publicKey;
            }
        }

        public static Ed25519Keypair New()
        {
            return new Ed25519Keypair(Native.newEd25519Keypair());
        }

        public static Ed25519Keypair FromPrivateKey(Ed25519PrivateKey privateKey)
        {
            return new Ed25519Keypair(Native.getEd25519KeypairFromPrivateKey(privateKey.Ptr));
        }

        public override void Dispose()
        {
            Native.destroyEd25519Keypair(Ptr);
        }
    }
}